flowmachine/flowmachine/features/subscriber/active_subscribers.py

```diff
@@ -0,0 +1,181 @@
+
+from flowmachine.core.mixins.exposed_datetime_mixin import ExposedDatetimeMixin
+from flowmachine.core.query import Query
+from flowmachine.features.subscriber.total_active_periods import (
+    TotalActivePeriodsSubscriber,
+)
+from flowmachine.features.utilities.events_tables_union import EventsTablesUnion
+import dateutil.rrule as rr
+from dateutil.relativedelta import relativedelta
+
+SubscriberSubsetType = NewType("SubscriberSubsetType", Union[str, list, Query, "Table"])
+
+
+class ActiveSubscribers(ExposedDatetimeMixin, Query):
+    """
+    Class that represents subscribers seen to be active.
+
+    To determine this, we regard the timeframe between `start_date` and `end_date` as the major period.
+    We then break the major period into `minor_period_count` minor periods.
+
+    A subscriber is considered to be active in a minor period if they are seen at least `minor_period_
+    threshold` times within that period.
+
+    A subscriber is considered active over the major period if they are active in at least `major_period_
+    threshold` periods.
+
+    Parameters
+    ----------
+    start_date, end_date: date, datetime, str
+        Major period between which to search for subscribers
+    minor_period_threshold: int
+        The number of times a subscriber must appear inside minor_period to count as active
+        in that period
+    major_period_threshold: int
+        The number of minor periods a subscriber must appear active in to appear in the output
+        of the query
+    subscriber_identifier : {'msisdn', 'imei'}, default 'msisdn'
+        Either msisdn, or imei, the column that identifies the subscriber.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    minor_period_count: int, default 24
+        The number of minor periods to split the major period into
+    minor_period_length: int, default 1
+        The number of period_units that make up a minor period
+    period_unit: {'days','hours','minutes'} default 'hours'
+        The unit of time to of minor_period_length
+
+    Notes
+    -----
+    * The date range will be inclusive os `start_date` but exclusive of `end_date` (ie range = start_
+    date ... (end_date - 1 second))
+    * The default values for minor_period_count and minor_period_length assume you wish to seach
+    for subscribers who are active at least `minor_period_threshold` hours throughout the day.
+
+    Examples
+    --------
+    Returns subscribers who were active on at least three hours between 2016-01-01 and 2016-01-02
+
+
+    >>>     active_subscribers = ActiveSubscribers(
+                start_date=date(year=2016, month=1, day=1),
+                end_date=date(year=2016, month=1, day=2),
```

Comment:
Personally I find this first example more confusing than helpful. This could be achieved without an `ActiveSubscribers` query, just using `TotalActivePeriodsSubscriber` with `numeric_subset`, and by counting subscribers active on 1 out of 1 days it's hard to see all the layers of the query. I'd suggest maybe getting rid of this one and just keeping the other two examples.